File: .understand-anything/.trash-1781514098969/tmp/generate.py
import json
import os

input_path = r"C:\SVN_CODE\branches\real\could_frontend\.understand-anything\tmp\ua-file-extract-results-35.json"
input_batch_path = r"C:\SVN_CODE\branches\real\could_frontend\.understand-anything\tmp\ua-file-analyzer-input-35.json"
out_dir = r"C:\SVN_CODE\branches\real\could_frontend\.understand-anything\intermediate"

with open(input_path, "r", encoding="utf-8") as f:
    extract_data = json.load(f)

with open(input_batch_path, "r", encoding="utf-8") as f:
    batch_data = json.load(f)

batch_import_data = batch_data.get("batchImportData", {})

nodes = []
edges = []

expert_knowledge = {
    "backend/apps/ai_models/migrations/0013_platform_llm_settings.py": {
        "summary": "AI模型应用的数据库迁移文件，负责清理旧的LLM模型绑定数据。",
        "tags": ["migration", "database", "ai-models"]
    },
    "backend/apps/ai_models/migrations/0014_remove_standalone_chat_room_menu.py": {
        "summary": "数据库迁移文件，用于从系统中移除独立的聊天室菜单项。",
        "tags": ["migration", "database", "menu-config"]
    },
    "backend/apps/ai_models/migrations/0015_tts_settings.py": {
        "summary": "TTS（文本转语音）设置的数据迁移文件，包含了从JSON文件加载并初始化默认语音数据的逻辑。",
        "tags": ["migration", "database", "tts", "data-seeding"]
    },
    "backend/apps/ai_models/tests/__init__.py": {
        "summary": "AI模型测试模块的初始化包文件。",
        "tags": ["test", "entry-point"]
    },
    "backend/apps/ai_models/tests/test_llm_provider_api.py": {
        "summary": "LLM提供商API的单元测试，主要验证已废弃的旧版接口是否被正确移除。",
        "tags": ["test", "api", "quality-assurance"]
    },
    "backend/apps/audit/__init__.py": {
        "summary": "审计日志(Audit)应用的初始化文件。",
        "tags": ["entry-point", "audit"]
    },
    "backend/apps/audit/apps.py": {
        "summary": "审计(Audit)应用的Django应用配置入口，定义了应用的元数据。",
        "tags": ["configuration", "django-app"]
    },
    "backend/apps/audit/migrations/__init__.py": {
        "summary": "审计模块迁移文件的包初始化文件。",
        "tags": ["entry-point", "migration"]
    },
    "backend/apps/audit/migrations/0001_initial.py": {
        "summary": "审计日志操作(OperationLog)数据表的初始创建迁移脚本。",
        "tags": ["migration", "database", "audit"]
    },
    "backend/apps/audit/migrations/0002_operationlog_description.py": {
        "summary": "审计日志数据表迁移文件，为操作日志新增描述(description)字段。",
        "tags": ["migration", "database", "audit"]
    },
    "backend/apps/audit/migrations/0003_operationlog_actor_role_name.py": {
        "summary": "审计日志数据表迁移文件，添加操作者角色名称字段以便更好地追踪行为。",
        "tags": ["migration", "database", "audit"]
    },
    "backend/apps/audit/migrations/0004_operationlog_actor_display_name.py": {
        "summary": "审计日志数据表迁移文件，添加操作者显示名称字段。",
        "tags": ["migration", "database", "audit"]
    },
    "backend/apps/audit/migrations/0005_detach_superuser_logs_and_fix_review_descriptions.py": {
        "summary": "审计日志数据修正迁移文件，分离超级管理员的日志记录，并修复知识库审核日志的描述文本。",
        "tags": ["migration", "database", "data-fix"]
    },
    "backend/apps/audit/tests/__init__.py": {
        "summary": "审计应用测试模块的初始化包文件。",
        "tags": ["test", "entry-point"]
    },
    "backend/apps/devices/__init__.py": {
        "summary": "设备管理(Devices)应用的初始化文件。",
        "tags": ["entry-point", "devices"]
    },
    "backend/apps/devices/AGENTS.md": {
        "summary": "设备管理模块的开发指南文档，详细记录了设备接入、鉴权流程及API设计规范。",
        "tags": ["documentation", "development", "architecture"]
    },
    "backend/apps/devices/apps.py": {
        "summary": "设备管理(Devices)应用的Django应用配置文件。",
        "tags": ["configuration", "django-app"]
    },
    "backend/apps/devices/management/__init__.py": {
        "summary": "设备管理应用的自定义命令模块初始化文件。",
        "tags": ["entry-point", "management"]
    },
    "backend/apps/devices/management/commands/__init__.py": {
        "summary": "设备管理自定义命令脚本的包初始化文件。",
        "tags": ["entry-point", "commands"]
    },
    "backend/apps/devices/management/commands/seed_operations_periodic_tasks.py": {
        "summary": "自定义Django管理命令，用于在系统中播种（初始化）设备运营相关的Celery周期性任务。",
        "tags": ["script", "utility", "celery", "data-seeding"]
    },
    "backend/apps/devices/migrations/__init__.py": {
        "summary": "设备管理迁移模块的初始化包文件。",
        "tags": ["entry-point", "migration"]
    },
    "backend/apps/devices/migrations/0001_initial.py": {
        "summary": "设备数据表的初始创建迁移脚本。",
        "tags": ["migration", "database", "devices"]
    },
    "backend/apps/devices/migrations/0002_device_tenant_alter_device_code_and_more.py": {
        "summary": "设备数据表迁移文件，为设备添加租户(tenant)关联并修改设备编码等字段属性。",
        "tags": ["migration", "database", "devices"]
    },
    "backend/apps/devices/migrations/0003_device_authorization_type_device_device_info_and_more.py": {
        "summary": "设备数据表的大型迁移文件，添加设备授权类型、设备详细信息JSON字段等多项属性，支持安卓等设备的免密登录体系。",
        "tags": ["migration", "database", "devices"]
    },
    "backend/apps/devices/migrations/0004_device_status_online_offline_only.py": {
        "summary": "设备数据表迁移文件，将设备状态精简为仅包含在线(online)与离线(offline)两种状态。",
        "tags": ["migration", "database", "devices"]
    }
}

for res in extract_data.get("results", []):
    path = res["path"]
    cat = res["fileCategory"]
    lines = res["nonEmptyLines"]
    
    node_type = "file"
    if cat == "infra":
        node_type = "service"
    elif cat == "data":
        node_type = "table"
    elif cat == "docs":
        node_type = "document"
    elif cat == "config":
        node_type = "config"
        
    name = os.path.basename(path)
    
    complexity = "simple"
    if lines > 200:
        complexity = "complex"
    elif lines >= 50:
        complexity = "moderate"
        
    knowledge = expert_knowledge.get(path, {})
    summary = knowledge.get("summary", "基础文件实现。")
    tags = knowledge.get("tags", ["utility"])
    
    lang_notes = None
    if "migration" in path.lower():
        lang_notes = "Django database migration module."
    
    nodes.append({
        "id": f"{node_type}:{path}",
        "type": node_type,
        "name": name,
        "filePath": path,
        "summary": summary,
        "tags": tags[:5],
        "complexity": complexity,
        "languageNotes": lang_notes
    })
    
    if "functions" in res:
        for fn in res["functions"]:
            fn_lines = fn.get("endLine", 0) - fn.get("startLine", 0) + 1
            is_exported = any(e["name"] == fn["name"] for e in res.get("exports", []))
            if fn_lines >= 10 or is_exported:
                fn_id = f"function:{path}:{fn['name']}"
                nodes.append({
                    "id": fn_id,
                    "type": "function",
                    "name": fn["name"],
                    "filePath": path,
                    "lineRange": [fn.get("startLine", 0), fn.get("endLine", 0)],
                    "summary": f"{fn['name']} 的具体业务逻辑实现。",
                    "tags": ["function", "utility"],
                    "complexity": "simple" if fn_lines < 20 else ("moderate" if fn_lines < 50 else "complex")
                })
                edges.append({
                    "source": f"{node_type}:{path}",
                    "target": fn_id,
                    "type": "contains",
                    "direction": "forward",
                    "weight": 1.0
                })
                if is_exported:
                    edges.append({
                        "source": f"{node_type}:{path}",
                        "target": fn_id,
                        "type": "exports",
                        "direction": "forward",
                        "weight": 0.8
                    })
                    
    if "classes" in res:
        for cls in res["classes"]:
            cls_lines = cls.get("endLine", 0) - cls.get("startLine", 0) + 1
            methods_count = len(cls.get("methods", []))
            is_exported = any(e["name"] == cls["name"] for e in res.get("exports", []))
            if cls_lines >= 20 or methods_count >= 2 or is_exported:
                cls_id = f"class:{path}:{cls['name']}"
                nodes.append({
                    "id": cls_id,
                    "type": "class",
                    "name": cls["name"],
                    "filePath": path,
                    "lineRange": [cls.get("startLine", 0), cls.get("endLine", 0)],
                    "summary": f"定义 {cls['name']} 类的数据结构或行为实现。",
                    "tags": ["class", "data-model"] if "Migration" not in cls["name"] else ["class", "migration"],
                    "complexity": "simple" if cls_lines < 50 else ("moderate" if cls_lines < 150 else "complex")
                })
                edges.append({
                    "source": f"{node_type}:{path}",
                    "target": cls_id,
                    "type": "contains",
                    "direction": "forward",
                    "weight": 1.0
                })
                if is_exported:
                    edges.append({
                        "source": f"{node_type}:{path}",
                        "target": cls_id,
                        "type": "exports",
                        "direction": "forward",
                        "weight": 0.8
                    })
                    
    imports = batch_import_data.get(path, [])
    for imp in imports:
        edges.append({
            "source": f"{node_type}:{path}",
            "target": f"file:{imp}",
            "type": "imports",
            "direction": "forward",
            "weight": 0.7
        })

print(f"Total Nodes: {len(nodes)}, Total Edges: {len(edges)}")

out_file = os.path.join(out_dir, "batch-35.json")
with open(out_file, "w", encoding="utf-8") as f:
    json.dump({"nodes": nodes, "edges": edges}, f, ensure_ascii=False, indent=2)

File: generate.js
const fs = require('fs');

const expertData = {
    // Batch 9
    "backend/apps/audit/admin.py": {
        summary: "配置 Django Admin 以管理和查看 OperationLog 审计日志模型。",
        tags: ["configuration", "admin", "audit-log"],
    },
    "backend/apps/audit/descriptions.py": {
        summary: "提供辅助函数，根据请求的方法、路径和载荷生成易于理解的中文操作描述。",
        tags: ["utility", "audit-log", "formatting"],
        functions: {
            "describe_operation": {
                summary: "根据 HTTP 请求详情映射并生成特定的业务操作中文描述。",
                tags: ["utility", "formatting"]
            }
        }
    },
    "backend/apps/audit/models.py": {
        summary: "定义 OperationLog 数据库模型，用于记录用户的各类操作行为、IP 地址、租户信息及请求详情。",
        tags: ["data-model", "audit-log", "database"],
        classes: {
            "OperationLog": {
                summary: "记录用户操作详情的数据库模型，包含租户关联、执行用户及请求路径等元数据。",
                tags: ["data-model", "database"]
            }
        }
    },
    "backend/apps/audit/serializers.py": {
        summary: "提供 OperationLog 的 DRF 序列化器，用于在 API 响应中格式化审计日志并嵌套展示操作者信息。",
        tags: ["serialization", "audit-log", "api-handler"],
        classes: {
            "OperationLogSerializer": {
                summary: "将审计日志模型实例转换为 JSON 数据，包含关联的用户详情解析。",
                tags: ["serialization"]
            }
        }
    },
    "backend/apps/audit/tests/test_operation_log_api.py": {
        summary: "审计日志模块的集成测试用例，覆盖日志查询、权限控制、租户数据隔离及日志清空等核心功能。",
        tags: ["test", "audit-log", "api-handler"],
        classes: {
            "OperationLogAPITests": {
                summary: "验证审计日志 API 端点的核心测试套件，确保权限机制和数据过滤正确工作。",
                tags: ["test"]
            }
        }
    },
    "backend/apps/audit/urls.py": {
        summary: "注册并路由审计日志相关的 DRF ViewSet，提供 /audit-logs/ 访问端点。",
        tags: ["entry-point", "routing", "audit-log"]
    },
    "backend/apps/audit/views.py": {
        summary: "处理审计日志的 API 视图集，支持按租户查询日志及具备特定权限时清空日志的高级操作。",
        tags: ["api-handler", "audit-log", "controller"],
        classes: {
            "OperationLogViewSet": {
                summary: "提供只读的日志查询接口及清空功能的视图集，内置多租户数据隔离逻辑。",
                tags: ["api-handler", "controller"]
            }
        }
    },
    
    // Batch 10
    "backend/config/sentry.py": {
        summary: "Sentry SDK 的初始化和配置逻辑，包含忽略特定健康检查路径和静态资源的过滤机制。",
        tags: ["configuration", "monitoring", "error-tracking"],
        functions: {
            "filter_transactions": {
                summary: "过滤请求事务，防止静态文件和内部健康检查路径被错误上报到 Sentry。",
                tags: ["monitoring", "filtering"]
            },
            "init_sentry": {
                summary: "初始化 Sentry SDK 并应用忽略规则、环境信息及性能采样率配置。",
                tags: ["configuration", "monitoring"]
            }
        }
    },
    "backend/config/settings/base.py": {
        summary: "Django 项目的基础配置文件，定义了应用注册、中间件、数据库连接、缓存及日志配置。",
        tags: ["configuration", "django", "core-settings"]
    },
    "backend/config/settings/dev.py": {
        summary: "本地开发环境专用的 Django 配置，开启调试模式并配置开发工具。",
        tags: ["configuration", "development", "django"]
    },
    "backend/config/settings/prod.py": {
        summary: "生产环境专用的 Django 配置，关闭调试模式并应用安全相关的生产策略。",
        tags: ["configuration", "production", "django"]
    },
    "backend/config/test_sentry_filters.py": {
        summary: "针对 Sentry 事件与事务过滤逻辑的测试文件，确保无效的上报请求被正确拦截。",
        tags: ["test", "monitoring", "error-tracking"],
        classes: {
            "SentryFilterTests": {
                summary: "测试 filter_transactions 等方法是否按预期屏蔽特定路由的性能追踪。",
                tags: ["test"]
            }
        }
    },

    // Batch 11
    "backend/config/__init__.py": {
        summary: "项目配置包的初始化入口，确保在 Django 启动时预先加载并注册 Celery 实例。",
        tags: ["entry-point", "configuration", "django"]
    },
    "backend/config/celery.py": {
        summary: "Celery 应用实例化及配置，负责从 Django settings 加载异步任务队列的参数配置。",
        tags: ["configuration", "task-queue", "async"]
    },
    "backend/config/tasks.py": {
        summary: "存放跨模块或通用的 Celery 异步任务，例如简单的调试探测任务。",
        tags: ["task-queue", "utility", "background-job"],
        functions: {
            "debug_task": {
                summary: "打印当前请求信息的调试任务，用于验证 Celery worker 进程的联通性。",
                tags: ["task-queue", "utility"]
            }
        }
    },

    // Batch 12
    "web/src/api/modules/tts.ts": {
        summary: "封装与语音合成 (TTS) 相关的 HTTP 接口调用逻辑，提供清晰的前端 API 访问层。",
        tags: ["api-handler", "tts", "integration"]
    },
    "web/src/views/tts-management/index.tsx": {
        summary: "用于管理音色模型及相关资源的 React 视图页面组件。",
        tags: ["component", "view", "tts"]
    },
    "web/src/views/tts-settings/index.tsx": {
        summary: "提供全局或特定应用的语音合成参数设置页面，例如语速、音量及默认声音的调节。",
        tags: ["component", "view", "tts"]
    },

    // Batch 13
    "backend/Dockerfile": {
        summary: "构建包含 Django、DRF 和 Celery 运行环境的镜像，通过多阶段构建优化体积，并封装服务启动序列。",
        tags: ["containerization", "infrastructure", "django"],
        languageNotes: "容器内启动序列和健康检查由根目录的 docker-compose 编排，本 Dockerfile 主要固化 Python 系统依赖。"
    },

    // Batch 14
    "web/Dockerfile": {
        summary: "构建 React 开发环境的 Docker 镜像，用于在容器内通过 Node 运行 Vite 开发服务器以支持热更新。",
        tags: ["containerization", "infrastructure", "frontend"],
        languageNotes: "本镜像为开发态使用 (npm run dev) 设计，通过匿名数据卷挂载依赖以规避宿主 npm 安装的污染问题。"
    }
};

const inputData = JSON.parse(fs.readFileSync('.understand-anything/tmp/batch-input-7.json', 'utf8'));

[9, 10, 11, 12, 13, 14].forEach(batchIndex => {
    const input = inputData.find(b => b.batchIndex === batchIndex);
    const extracted = JSON.parse(fs.readFileSync('.understand-anything/tmp/ua-file-extract-results-' + batchIndex + '.json', 'utf8'));
    
    let nodes = [];
    let edges = [];

    extracted.results.forEach(file => {
        const ed = expertData[file.path] || { summary: "自动生成的文件节点。", tags: ["file"] };
        
        let nodeType = 'file';
        if (file.fileCategory === 'config') nodeType = 'config';
        if (file.fileCategory === 'infra') nodeType = file.path.includes('Dockerfile') ? 'service' : 'pipeline';
        if (file.fileCategory === 'data' && file.path.endsWith('.sql')) nodeType = 'table';
        
        nodes.push({
            id: `${nodeType}:${file.path}`,
            type: nodeType,
            name: file.path.split('/').pop(),
            filePath: file.path,
            summary: ed.summary,
            tags: ed.tags,
            complexity: file.totalLines > 200 ? 'complex' : (file.totalLines > 50 ? 'moderate' : 'simple'),
            languageNotes: ed.languageNotes || undefined
        });

        // Functions
        if (file.functions) {
            file.functions.forEach(fn => {
                let isExported = file.exports?.find(e => e.name === fn.name);
                if (fn.endLine - fn.startLine >= 10 || isExported) {
                    let fnNodeId = `function:${file.path}:${fn.name}`;
                    let fnEd = ed.functions && ed.functions[fn.name] ? ed.functions[fn.name] : {summary: `${fn.name} 的具体实现代码。`, tags: ["function"]};
                    nodes.push({
                        id: fnNodeId,
                        type: 'function',
                        name: fn.name,
                        filePath: file.path,
                        lineRange: [fn.startLine, fn.endLine],
                        summary: fnEd.summary,
                        tags: fnEd.tags,
                        complexity: (fn.endLine - fn.startLine) > 50 ? 'complex' : 'simple'
                    });
                    edges.push({
                        source: `${nodeType}:${file.path}`,
                        target: fnNodeId,
                        type: 'contains',
                        direction: 'forward',
                        weight: 1.0
                    });
                    if (isExported) {
                        edges.push({
                            source: `${nodeType}:${file.path}`,
                            target: fnNodeId,
                            type: 'exports',
                            direction: 'forward',
                            weight: 0.8
                        });
                    }
                }
            });
        }

        // Classes
        if (file.classes) {
            file.classes.forEach(cls => {
                let isExported = file.exports?.find(e => e.name === cls.name);
                if (cls.endLine - cls.startLine >= 20 || (cls.methods && cls.methods.length >= 2) || isExported) {
                    let clsNodeId = `class:${file.path}:${cls.name}`;
                    let clsEd = ed.classes && ed.classes[cls.name] ? ed.classes[cls.name] : {summary: `${cls.name} 的类定义。`, tags: ["class"]};
                    nodes.push({
                        id: clsNodeId,
                        type: 'class',
                        name: cls.name,
                        filePath: file.path,
                        lineRange: [cls.startLine, cls.endLine],
                        summary: clsEd.summary,
                        tags: clsEd.tags,
                        complexity: (cls.endLine - cls.startLine) > 100 ? 'complex' : 'moderate'
                    });
                    edges.push({
                        source: `${nodeType}:${file.path}`,
                        target: clsNodeId,
                        type: 'contains',
                        direction: 'forward',
                        weight: 1.0
                    });
                    if (isExported) {
                        edges.push({
                            source: `${nodeType}:${file.path}`,
                            target: clsNodeId,
                            type: 'exports',
                            direction: 'forward',
                            weight: 0.8
                        });
                    }
                }
            });
        }
        
        // Imports
        const imports = input.batchImportData[file.path] || [];
        imports.forEach(imp => {
            edges.push({
                source: `${nodeType}:${file.path}`,
                target: `file:${imp}`, // imports generally point to files unless infra
                type: 'imports',
                direction: 'forward',
                weight: 0.7
            });
        });
        
        // Extra edges from expertData
        if (ed.extraEdges) {
            edges.push(...ed.extraEdges.map(e => ({...e, source: `${nodeType}:${file.path}`})));
        }

        // Handle specific non-code relationships
        if (file.fileCategory === 'infra' && file.path.includes('Dockerfile')) {
            if (file.path.startsWith('backend/')) {
                edges.push({
                    source: `${nodeType}:${file.path}`,
                    target: `file:backend/config/settings/base.py`,
                    type: 'deploys',
                    direction: 'forward',
                    weight: 0.7
                });
            } else if (file.path.startsWith('web/')) {
                edges.push({
                    source: `${nodeType}:${file.path}`,
                    target: `file:web/package.json`, // though not strictly in batch, it represents the app
                    type: 'deploys',
                    direction: 'forward',
                    weight: 0.7
                });
            }
        }
    });

    fs.writeFileSync(`.understand-anything/intermediate/batch-${batchIndex}.json`, JSON.stringify({nodes, edges}, null, 2), 'utf8');
    console.log(`Wrote batch-${batchIndex}.json with ${nodes.length} nodes and ${edges.length} edges.`);
});